include/boost/geometry/algorithms/detail/check_iterator_range.hpp

```diff
@@ -59,10 +59,74 @@ struct check_iterator_range
                 return false;
             }
         }
+
         return AllowEmptyRange || first != beyond;
     }
 };
 
+// As check_iterator_range but the index of the element that
+// satisfies the predicate is returned otherwise -1 is returned
+template <typename Predicate>
+struct check_iterator_range_with_index
+{
+
+    // version where we can pass a predicate object
+    template <typename InputIterator>
+    static inline int apply(InputIterator first,
+                             InputIterator beyond,
+                             Predicate const& predicate)
+    {
+        // in case predicate's apply method is static, MSVC will
+        // complain that predicate is not used
+        boost::ignore_unused(predicate);
+
+        for (InputIterator it = first; it != beyond; ++it)
+        {
+            if (! predicate.apply(*it))
+            {
+                return it - first;
+            }
+        }
+
+        return -1;
+    }
+};
+
+
+// As check_iterator_range but the index of the element that
+// satisfies the predicate is returned otherwise -1 is returned
+template <typename Geometry, typename Construction>
+struct check_iterator_range_with_info
+{
+    typedef typename point_type<Geometry>::type point_type;
+
+    typedef geometry::segment_intersection_points<point_type>
+    intersection_return_type;
+
+    // version where we can pass a predicate object
+    template <typename InputIterator>
+    static inline intersection_return_type
+    apply(InputIterator first,
+          InputIterator beyond,
+          Construction const& construction)
+    {
+        // in case predicate's apply method is static, MSVC will
+        // complain that predicate is not used
+        boost::ignore_unused(construction);
+
+        for (InputIterator it = first; it != beyond; ++it)
+        {
+            intersection_return_type res = construction.apply(*it);
+            if (res.count != 0)
```

Comment:
Why it has to be here instead of being checked inside `predicate` called in `check_iterator_range_with_index` or `std::find_if`?

Reply:
`check_iterator_range_with_index` does not take a predicate but a construction (i.e.  return some non boolean result), I cannot see a much simpler way than the one I use here.

Reply:
Yes unless `check_iterator_range_with_index` took `predicate` as reference instead of const reference, internally called `construction.apply(*it)` and stored the last value inside, or if the predicate was taken by copy and returned from the function (like std::for_each but with stopping condition https://en.cppreference.com/w/cpp/algorithm/for_each), or the function returned `std::pair<index_type, predicate_type>`, etc.